### PR TITLE
fix: `super.x` in a loop

### LINKED ISSUE #15945: [Bug]: `plugin-transform-object-super` incorrect output where object created in a loop
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

```js
const objects = [];
for (const proto of [{x: 0}, {x: 1}]) {
  objects.push({
    __proto__: proto,
    foo() {
      return super.x
    }
  });
}
assert(objects[0].foo() === 0);
assert(objects[1].foo() === 1);
```

[REPL](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=MYewdgzgLgBCBGArApsKEYF4YG0C6A3AFABmIATjABSiSwAO5IUIcJuA3gB4BcMADAF8ANDG58AjILwBKMURhwkqdADp6AVwgALKhwWKYAfSONmIE3zMthBxWRBU5-w4fLIoG8mBgQN9ZHJVLjsYQQNBGWJwgEMICECoKgQUNAgcfjxVBycsTGx-KKI4hPIklJV0iSycuXzsCSKgA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=&prettier=false&targets=&version=7.22.14&externalPlugins=%40babel%2Fplugin-transform-object-super%407.18.6&assumptions=%7B%7D)

### Configuration file name

_No response_

### Configuration

_No response_

### Current and expected behavior

In the original code, the assertions pass. When transformed with [@babel/plugin-transform-object-super](https://babeljs.io/docs/babel-plugin-transform-object-super), the first assertion fails.

The transformed output is:

```js
const objects = [];
for (const proto of [{x: 0}, {x: 1}]) {
  var _obj;
  objects.push(
    (_obj = {
      __proto__: proto,
      foo() {
        return _get(_getPrototypeOf(_obj), "x", this);
      }
    })
  );
}
assert(objects[0].foo() === 0);
assert(objects[1].foo() === 1);
```

### Environment

REPL

### Possible solution

The problem is that `_obj` for the 1st turn of the loop is over-written in 2nd turn, because `_obj` is not block scoped. Could be solved by changing `var _obj` to `let _obj`.

Please see [REPL](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=MYewdgzgLgBCBGArApsKEYF4YG0C6A3AFABmIATjABSiSwAO5IUIcJuA3gB4BcMADAF8ANDG58AjILwBKMURhwkqdADp6AVwgALKhwWKYAfSONmIE3zMthBxWRBU5-w4fLIoG8mBgQN9ZHJVLjsYQQNBGWJwgEMICECoKgQUNAgcfjxVBycsTGx-KKI4hPIklJV0iSycuXzsCSKgA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=&prettier=false&targets=&version=7.22.14&externalPlugins=%40babel%2Fplugin-transform-object-super%407.18.6&assumptions=%7B%7D).

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by liuxingbaoyu**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #15945  
| Patch: Bug Fix?          | √
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | √
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15948"></a>

### CODE REVIEW

**Review by nicolo-ribaudo — APPROVED** [review]
I think it's ok to start emitting `let` for the loop case even if it's ES6, given that anybody using `super` will also be compiling `let`/`const` anyway. However, would it be possible to only emit `let` if we are actually in a loop, to reduce risk?

**Review by JLHwung — APPROVED** [review]

File: packages/babel-plugin-transform-object-super/src/index.ts [hunk 1]
```diff
@@ -37,7 +37,11 @@ export default declare(api => {
         });
 
         if (objectRef) {
-          path.scope.push({ id: t.cloneNode(objectRef) });
+          const scopePath = path.scope.getBlockParent().path;
+          path.scope.push({
+            id: t.cloneNode(objectRef),
+            kind: scopePath.parentPath?.isLoop() ? "let" : "var",
+          });
```
Comment (lightmare) [inline packages/babel-plugin-transform-object-super/src/index.ts:44]: Checking direct parent is not sufficient. For example, wrap the loop body in `if (true) { ... }` and you get a `var` again.

Comment (liuxingbaoyu) [reply packages/babel-plugin-transform-object-super/src/index.ts:46]: Fixed, thank you!

**Comment by lightmare** [comment]
Found another issue, probably related to plugin order. `plugin-transform-block-scoping` will simply rewrite the `let` to a `var`, unless it's already capturing the scope because of something else.

This fails, but if you comment-out the `bar()` method, it passes: [REPL](https://babeljs.io/repl/build/55412/#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=PTC04KAhjB7A7AzgFwASwEYCsCm1mKoC8qA2gLoDcUAZrAE6oAUcSaADvbMrOjWQG8AHgC5UABgC-AGlTCxARknkAlBFSoAlvybJ6AVxwq56jemx4CAOnb7EACyYDTZ1AH03nbrA9ivPaRczdj8uAKCNOlgmY2dXV3ocZH16eFREfXYceishCNQZCGAAKnyMAEN6GJN4s0Tk1NR_WFz8wpBi4CDJFWoNSVoGZlYUdD5zXHxEWNMR2AAbHCt52ABzJhb2XOIiEhaomNlN3KOrA5VeiAG6RhYEUd5YfkxJghmNOcWrcsREbOQNjZtrs9mdYNEVKctkJTudLgMgA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=true&sourceType=module&lineWrap=false&presets=&prettier=false&targets=&version=7.22.16%2Bpr.15948&externalPlugins=%40babel%2Fplugin-transform-block-scoping%407.21.0%2C%40babel%2Fplugin-transform-object-super%407.18.6&assumptions=%7B%7D)
```js
//---

const objects = [];

for (const proto of [{x: 0}, {x: 1}])
  if (true) {
    objects.push({
      __proto__: proto,
      p: proto,
      foo() {
        return super.x
      },
/*
      bar() {
        return proto.x
      },
//*/
    });
  }

for (const o of objects) {
  console.log(o.p.x === o.foo(), o.p.x, o.foo());
}
for (const o of objects) {
  console.assert(o.p.x === o.foo(), o.p.x, o.foo());
}
```

**Comment by liuxingbaoyu** [comment]
Amazing find, thank you!
This is actually a bug in `plugin-transform-block-scoping`, it's probably a regression, I'm surprised no one has reported it. :)
I will investigate.

[old babel](https://bvaughn.github.io/babel-repl/#?babili=false&browsers=&builtIns=false&code_lz=G4QwTgBA9gRgVgUwMYBcDOEC8EDaBdAbgChRIBLLCABmIDMpIAKUiABzChSmltyoBoIARgj8ATHgCUEMr0YowAVwTSA3kQgQANghQQA-rDiUyxTWQDUFsxA3R4ydADpWitAAtG6zZvpRGanY-EGC6imAAdgZGNpoAvnZxksQJREhQEWhQOk4gaGgIYCiMRo5oOFR4Tn4BmHUCpajlldVQ_pLJRERAA&debug=false&evaluate=false&lineWrap=true&presets=es2015%2Creact%2Cstage-2&prettier=false&showSidebar=true&targets=)

```js
var objects = [];
var i = 0;
for (var proto of [0, 1 ,2]) if (true) {
  let _obj = i;
  i++;
  
  objects.push({
    foo() {
      return _obj;
    }
  });
}

console.assert(objects[0].foo()===0,objects[0].foo());

```


[repl](https://babeljs.io/repl/build/55412/#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=G4QwTgBA9gRgVgUwMYBcDOEC8EDaBdAbgChRIBLLCABmIDMpIAKUiABzChSmltyoBoIARgj8ATHgCUEMr0YowAVwTSA3kQgQANghQQA-rDiUyxTWQDUFsxA3R4ydADpWitAAtG6zZvpRGanY-EGC6imAAdgZGNpoAvnZxksQJREhQEWhQOk4gaGgIYCiMRo5oOFR4Tn4BmHUCpajlldVQ_pLJRERAA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=true&sourceType=module&lineWrap=false&presets=&prettier=false&targets=&version=7.22.16%2Bpr.15948&externalPlugins=%40babel%2Fplugin-transform-block-scoping%407.21.0&assumptions=%7B%7D)

**Comment by nicolo-ribaudo** [comment]
Awesome, thanks @lightmare for the review :)

@liuxingbayou Do you mind splitting [8eda7f0](https://github.com/babel/babel/pull/15948/commits/8eda7f00356bd6d3179e575a473f065feed02532) into a separate PR so that we have to separate changelog entries? 🙏

**Comment by nicolo-ribaudo** [comment]
Could you rebase? :)

**Comment by nicolo-ribaudo** [comment]
I opened https://github.com/babel/babel/issues/16058 to keep track of the other regression.

**Pull request merged**